### DISCUSSION TITLE
Welcome to wrapit Discussions!

### DISCUSSION THREAD

**Discussion by grasph** [grasph/wrapit#49] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by rajkumardongre** [comment]
Hi everyone! I'm Rajkumar, a Bachelor of Technology student specializing in Artificial Intelligence and Data Science. I'm excited to be here and connect with the community!

My programming experience spans Fortran, C, C++, Java, Python, and Julia. I even had the amazing opportunity to contribute to the Fortran-lang organization's HTTP client package during Google Summer of Code 2023, enabling HTTP requests in Fortran.

For Google Summer of Code 2024, I'm particularly interested in the **"Julia Interoperating with HEP C++ Libraries"** project. I'm eager to leverage my skills and enthusiasm to contribute to this valuable initiative.

I'm looking forward to learning, engaging, and collaborating with all of you in this wonderful community!

**Comment by rajkumardongre** [comment]
**Title:** Correct way to store strings/array in ROOT files with Julia

Hello everyone,  I'm trying to store string data in a ROOT file. I came across the `write_tree1.jl` example(in `examples/ex002-ROOT/TTree_examples/write_tree1.jl`), where vector length (`brCnt`) is stored alongside float vectors (`brMuon_pt`, `brMuon_eta`, `brMuon_phi`). 

My question:

* Can we apply the same approach to store strings (as they are essentially character arrays) by adding a separate branch for string length?
* Are there other ways to store strings or arrays in ROOT files with Julia, without the need to store the length separately?

Here's a simple example illustrating the approach I'm considering:

```julia
# Program to store strings in root file 
using ROOT

f = ROOT.TFile!Open("string.root", "RECREATE")
t = ROOT.TTree("tree", "tree")

a = "xyz"
n = fill(length(a))

Branch(t, "n", n, 32000, 99)
Branch(t, "string[n]", a, 32000, 99)
Fill(t)

Write(t)
Close(f)
```

I'd appreciate any insights or suggestions from the community on the best approach for storing strings effectively in ROOT files using Julia. 

Thank you

**Reply by grasph** [reply]
The `Branch` method version used in the [`ex002-ROOT/TTree_examples`](https://github.com/grasph/wrapit/tree/main/examples/ex002-ROOT/TTree_examples) examples does not support the `String` type. Supported types can be found in the file [`ex002-ROOT/Templates.h`](https://github.com/grasph/wrapit/tree/main/examples/ex002-ROOT/TTree_examples).

To store strings in a `TTree` using this `ROOT` wrapper, the method you proposed can be used. There exist two other options. 

### Option 1. Add string support to the `Branch` method used in the examples

Adding in `Template.h` function template instantiations for the `std::string` type should provide support for strings stored as `std::string` in the `TTree`. Saving them in the `TTree` as c-string type can be achieved by implementing specialized methods that will handle this specific case: need to call the `TTree:Branch()` function version mentioned in the next option.

### Option 2. Use a different `Branch` method.

Strings can already be saved as c-strings with the current wrapper code, using the Julia binding of [this version of the Branch method [2/13]](https://root.cern/doc/v630/classTTree.html#ad10fdaab73d4e756aa785d38b1b1f8d8), at the price of some pointer gymnastics. Here is an example of how to use this method:


```julia
using ROOT

tree = ROOT.TTree("tree", "tree")
br = Branch(tree, "s", Ptr{Nothing}(), "s/C")
s  = "ABC"
GC.@preserve s begin
    SetBranchAddress(tree, "s", convert(Ptr{Int8}, pointer(s)))
    Fill(tree)
end

#Show the data stored in the tree
Scan(tree)
```

**Reply by rajkumardongre** [reply]
Thanks, Philippe, I'll give this a try.

**Comment by rajkumardongre** [comment]
While attempting to address the first TODO item in the [WrapIt development TODO list](https://github.com/grasph/wrapit/blob/main/src/TODO.md), which involves fixing issues indicated with `test/TestTemplate3` and `test/TestTemplate4`, While executing all tests, I encountered a failure in the `TestCtorDefVal` test. Here's the sequence of events:

1. I initiated the test suite using the command:
   ```
   cd test && julia runtests.jl
   ```

2. The `TestSizet` test passed successfully, but the `TestCtorDefVal` test failed with the following error:
   ```
   julia runtests.jl TestCtorDefVal
   ```

**Error Message:**

```
Activating project at `~/Desktop/Coding/CERN/wrapit/test`
Try to run cmake...
  Activating project at `~/Desktop/Coding/CERN/wrapit/buildtools`
-- Found Julia executable: /home/rajkumar/.juliaup/bin/julia
-- Julia_VERSION_STRING: 1.10.2
-- Julia_INCLUDE_DIRS:   /home/rajkumar/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/include/julia
-- Julia_LIBRARY_DIR:    /home/rajkumar/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/lib
-- Julia_LIBRARY:        /home/rajkumar/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/lib/libjulia.so.1.10
-- JULIA_HOME:           /home/rajkumar/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/bin
-- Julia_LLVM_VERSION:   v15.0.7
-- Julia_WORD_SIZE:      64
-- Found JlCxx at /home/rajkumar/.julia/artifacts/5209ca23f516fb3391b885eef717e49b4ee0a268/lib
Found wrapit executable: /usr/local/bin/wrapit
'/usr/local/bin/wrapit' '-v' '1' '--force' '--update' '--cmake' '--output-prefix' '/home/rajkumar/Desktop/Coding/CERN/wrapit/test/TestCtorDefVal/build' '/home/rajkumar/Desktop/Coding/CERN/wrapit/test/TestCtorDefVal/TestCtorDefVal.wit'
The configuration file misses the uuid parameter. Folloing generated uuid will be used for the generate Julia project. Add the following line in the .wit configuration file to use same uuid for next versions of the code.
uuid = "228b8ce3-e52d-147c-62e2-402ca10cfb02"

Clang options:
        -resource-dir
        /usr/lib/llvm-13/lib/clang/13.0.1
        -x
        c++-header
        -std=c++17
        -I
        .
        -D
        WRAPIT
        -MMD
        -MF/home/rajkumar/Desktop/Coding/CERN/wrapit/test/TestCtorDefVal/build/wrapit.cmake

Info: Generating non-const getter for A::i_
Info: Generating non-const getter for A::i_

Generated wrapper statictics
  enums:            0
  classes/structs:  1
    templates:      0
    others:         1
  class methods:    1
  field accessors:  1 getters and 1 setters
  global variable accessors: 0 getters and 0 setters
  global functions: 0

-- Configuring done (6.1s)
-- Generating done (0.0s)
-- Build files have been written to: /home/rajkumar/Desktop/Coding/CERN/wrapit/test/TestCtorDefVal/build
[ 50%] Building CXX object CMakeFiles/jlTestCtorDefVal.dir/libTestCtorDefVal/src/jlTestCtorDefVal.cxx.o
In file included from /home/rajkumar/.julia/artifacts/5209ca23f516fb3391b885eef717e49b4ee0a268/include/jlcxx/module.hpp:14,
                 from /home/rajkumar/.julia/artifacts/5209ca23f516fb3391b885eef717e49b4ee0a268/include/jlcxx/jlcxx.hpp:15,
                 from /home/rajkumar/Desktop/Coding/CERN/wrapit/test/TestCtorDefVal/build/libTestCtorDefVal/src/jlTestCtorDefVal.cxx:2:
/home/rajkumar/.julia/artifacts/5209ca23f516fb3391b885eef717e49b4ee0a268/include/jlcxx/attr.hpp: In instantiation of ‘void jlcxx::detail::parse_attributes_helper(ExtraFunctionData&, T) [with T = bool]’:
/home/rajkumar/.julia/artifacts/5209ca23f516fb3391b885eef717e49b4ee0a268/include/jlcxx/attr.hpp:156:29:   required from ‘jlcxx::detail::ExtraFunctionData jlcxx::detail::parse_attributes(Extra ...) [with bool AllowCallingPolicy = false; bool AllowFinalizePolicy = true; Extra = {bool}]’
/home/rajkumar/.julia/artifacts/5209ca23f516fb3391b885eef717e49b4ee0a268/include/jlcxx/module.hpp:623:79:   required from ‘void jlcxx::Module::constructor(jl_datatype_t*, Extra ...) [with T = A; ArgsT = {}; Extra = {bool}; jl_datatype_t = _jl_datatype_t]’
/home/rajkumar/.julia/artifacts/5209ca23f516fb3391b885eef717e49b4ee0a268/include/jlcxx/module.hpp:1075:40:   required from ‘jlcxx::TypeWrapper<T>& jlcxx::TypeWrapper<T>::constructor(Extra ...) [with ArgsT = {}; Extra = {bool}; T = A]’
/home/rajkumar/Desktop/Coding/CERN/wrapit/test/TestCtorDefVal/build/libTestCtorDefVal/src/jlTestCtorDefVal.cxx:28:29:   required from here
/home/rajkumar/.julia/artifacts/5209ca23f516fb3391b885eef717e49b4ee0a268/include/jlcxx/attr.hpp:139:31: error: incomplete type ‘jlcxx::detail::process_attribute<bool, void>’ used in nested name specifier
  139 |     process_attribute<T>::init(std::forward<T_>(argi), f);
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
gmake[2]: *** [CMakeFiles/jlTestCtorDefVal.dir/build.make:76: CMakeFiles/jlTestCtorDefVal.dir/libTestCtorDefVal/src/jlTestCtorDefVal.cxx.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/jlTestCtorDefVal.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
TestCtorDefVal: Test Failed at /home/rajkumar/Desktop/Coding/CERN/wrapit/test/runtests.jl:36
  Expression: begin
    #= /home/rajkumar/Desktop/Coding/CERN/wrapit/test/runtests.jl:37 =#
    try
        #= /home/rajkumar/Desktop/Coding/CERN/wrapit/test/runtests.jl:38 =#
        if isfile(joinpath(source_dir, "CMakeLists.txt"))
            #= /home/rajkumar/Desktop/Coding/CERN/wrapit/test/runtests.jl:39 =#
            println("Try to run cmake...")
            #= /home/rajkumar/Desktop/Coding/CERN/wrapit/test/runtests.jl:42 =#
            run(`cmake -S $source_dir -B $build_dir`)
            #= /home/rajkumar/Desktop/Coding/CERN/wrapit/test/runtests.jl:43 =#
            run(`cmake --build $build_dir -j $ncores`)
        else
            #= /home/rajkumar/Desktop/Coding/CERN/wrapit/test/runtests.jl:45 =#
            println("source_dir:", source_dir)
            #= /home/rajkumar/Desktop/Coding/CERN/wrapit/test/runtests.jl:47 =#
            run(`make -C $source_dir -j $ncores`)
        end
        #= /home/rajkumar/Desktop/Coding/CERN/wrapit/test/runtests.jl:49 =#
        true
    catch
        #= /home/rajkumar/Desktop/Coding/CERN/wrapit/test/runtests.jl:51 =#
        false
    end
end

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] macro expansion
   @ ~/Desktop/Coding/CERN/wrapit/test/runtests.jl:36 [inlined]
 [3] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [4] macro expansion
   @ ~/Desktop/Coding/CERN/wrapit/test/runtests.jl:33 [inlined]
 [5] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [6] top-level scope
   @ ~/Desktop/Coding/CERN/wrapit/test/runtests.jl:31

Fail-fast enabled: Fail or Error occurred

Test Summary:    | Fail  Total   Time
Tests            |    1      1  10.7s
  TestCtorDefVal |    1      1  10.6s
ERROR: LoadError: Some tests did not pass: 0 passed, 1 failed, 0 errored, 0 broken.
in expression starting at /home/rajkumar/Desktop/Coding/CERN/wrapit/test/runtests.jl:30
```

**Steps to Reproduce:**

1. Navigate to the `test` directory.
2. Run the tests using the command: `julia runtests.jl`.
3. Observe the failure in the `TestCtorDefVal` test.

**Additional Information:**

- Julia Version: 1.10.2
- Operating System: Ubuntu 23.10

**Reply by grasph** [reply]
Thanks @rajkumardongre for the report. The problem comes from a CxxWrap update. I've filled an [issue](https://github.com/grasph/wrapit/issues/51) . I will work on fixing it next week. Until it is fixed, you can use CxxWrap v0.14.2: from the package manage: `add CxxWrap@v0.14.2`, then `pin CxxWrap` to prevent further updates. The `runtests.jl` script use its folders as project directory. So, for the tests, the operation must be done in a Julia session started from the `test` directory with the `--project` option.

Philippe.

**Reply by rajkumardongre** [reply]
Thanks @grasph for the update. While all of the tests are passing successfully, I encountered an error specifically with the `ex002-ROOT` test. Here's the error message:

```
...
...
LD_LIBRARY_PATH=libROOT:/home/rajkumar/root/root-6.30.04-install/lib JULIA_LOAD_PATH=:`pwd`/build/ROOT/src julia demo_ROOT.jl
Warning in <TCanvas::Delete()>: may not use this method
****************************************
Minimizer is Minuit2 / Migrad
Chi2                      =      57.2152
NDf                       =           66
Edm                       =  9.31601e-06
NCalls                    =           55
Constant                  =      197.704   +/-   3.46748     
Mean                      =    0.0172556   +/-   0.014353    
Sigma                     =      0.99873   +/-   0.0103909     (limited)
Info in <TCanvas::Print>: file demo_ROOT.png has been created
cmp demo_ROOT.png demo_ROOT-ref.png
demo_ROOT.png demo_ROOT-ref.png differ: byte 64, line 3
make: *** [Makefile:76: test] Error 1
ex002-ROOT: Test Failed at /home/rajkumar/Desktop/Coding/CERN/wrapit/test/test_examples.jl:52
  Expression: run_ex002_ROOT

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] macro expansion
   @ ~/Desktop/Coding/CERN/wrapit/test/test_examples.jl:52 [inlined]
 [3] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [4] top-level scope
   @ ~/Desktop/Coding/CERN/wrapit/test/test_examples.jl:43
Test Summary: | Fail  Total      Time
ex002-ROOT    |    1      1  11m49.2s
ERROR: LoadError: Some tests did not pass: 0 passed, 1 failed, 0 errored, 0 broken.
in expression starting at /home/rajkumar/Desktop/Coding/CERN/wrapit/test/test_examples.jl:40
```

**Steps Used:**

1. Cloned the repository:
   ```
   $ git clone https://github.com/grasph/wrapit.git
   $ cd wrapit/test
   ```

2. Launched Julia session with the project directory:
   ```
   $ julia --project=.
   ```

3. Within Julia's package mode, included the test_examples.jl file:
   ```julia
   julia> ]
   (test) pkg> include("test_examples.jl")
   ```

**Reply by grasph** [reply]
Hello Rajkumar,

As we can see in the error message a bit-by-bit comparison of two images is done to validate the test (see line `cmp demo_ROOT.png demo_ROOT-ref.png`): the test checks that the demo example works. If the `demo_ROOT.png` file contains a plot (like the reference one, `demo_ROOT-ref.png`), then it was a false-positive (or false-negative depending on the point of view) and the test is actually ok. Otherwise, it means there is some issue with running the ROOT demo,

Philippe.
